### PR TITLE
improved cancel and empty stack behaviour

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- introduced new stack frame type
- introduced `CLEANUP` step at the end of every flow 
- increased circuit breaker threshold to better support corrections
- removed unecessary `RESUME` flow stack type
- added pattern that gets triggered when the stack runs empty

Fixes https://rasahq.atlassian.net/browse/ENG-476
Fixes https://rasahq.atlassian.net/browse/ENG-462
Fixes https://rasahq.atlassian.net/browse/ENG-468

### CODE REVIEW

File: rasa/core/policies/flow_policy.py [hunk 1]
```diff
@@ -721,6 +721,16 @@ def _run_step(
                         # TODO: we need to figure out how to actually "undo" the
                         #    changed slots
                         pass
+                elif not previous_flow and current_frame.flow_id != "pattern_completed":
+                    # TODO: rather than triggering the flow-trigger-action this should
+                    #   directly put the flow onto the stack and start it that way.
+                    return ActionPrediction(
+                        FLOW_PREFIX + "pattern_completed",
+                        1.0,
+                        metadata={"slots": {PREVIOUS_FLOW_SLOT: previous_flow_name}},
```
Comment (twerkmeister) [inline rasa/core/policies/flow_policy.py:730]: won't this previous flow name always be None here? Ah you just resetting that slot, right? Might be more clear if you just set it to None

**Review by twerkmeister — APPROVED** [review]
Looks good, some suggestions and comments

File: rasa/cdu/flow_stack.py [hunk 2]
```diff
@@ -146,14 +146,16 @@ class StackFrameType(str, Enum):
     """The frame is an interrupt frame.
 
     This means that the previous flow was interrupted by this flow."""
+    REMARK = "remark"
+    """The frame is a short remark - in a sense a short interruption.
+
+    This means that the previous flow was interrupted by this flow, but the
+    interruption is short and the previous flow should be resumed after this
+    flow is finished."""
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/flow_stack.py:154]: yes

File: rasa/core/policies/flow_policy.py [hunk 3]
```diff
@@ -685,18 +685,18 @@ def _run_step(
...
+                # be continued now that this one has ended.
+                previous_flow_name = (
+                    previous_flow.name or previous_flow.id if previous_flow else None
+                )
+
                 if (
                     current_frame.frame_type == StackFrameType.INTERRUPT
                     and previous_flow is not None
+                    and previous_flow_step is not None
+                    and previous_flow_step.id != END_STEP
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/policies/flow_policy.py:698]: yeah saw this happening 😂

File: rasa/core/processor.py [hunk 4]
```diff
@@ -79,7 +79,7 @@
 logger = logging.getLogger(__name__)
 structlogger = structlog.get_logger()
 
-MAX_NUMBER_OF_PREDICTIONS = int(os.environ.get("MAX_NUMBER_OF_PREDICTIONS", "10"))
+MAX_NUMBER_OF_PREDICTIONS = int(os.environ.get("MAX_NUMBER_OF_PREDICTIONS", "20"))
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/processor.py:82]: I know, ideally - we find a different approach as part of https://rasahq.atlassian.net/browse/ENG-488

File: rasa/shared/core/flows/flow.py [hunk 5]
```diff
@@ -504,7 +504,14 @@ def __init__(self) -> None:
...
             metadata={},
-            next=FlowLinks(links=[]),
+            # The end step links to itself. This is needed to make sure that
+            # this allows us to end a flow by setting the active step of a flow
+            # to the end step.
+            # Since the side effects of a node are executed on the transition
+            # to the next node, we need this link to run the END logic.
+            # Otherwise, setting a flow to its end step would not execute the
+            # side effects of the end step.
+            next=FlowLinks(links=[StaticFlowLink(target=END_STEP)]),
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/shared/core/flows/flow.py:514]: Yes I thought about adding a `WRAPUP` step after `END`, wdyt?

**Pull request merged**